src/AI/Muriel.java

```diff
@@ -0,0 +1,750 @@
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Queue;
+import java.util.Set;
+
+import CommandingOfficers.Commander;
+import CommandingOfficers.CommanderAbility;
+import Engine.GameAction;
+import Engine.GameAction.ActionType;
+import Engine.GameActionSet;
+import Engine.Utils;
+import Engine.XYCoord;
+import Engine.Combat.BattleInstance;
+import Terrain.Environment;
+import Terrain.Environment.Weathers;
+import Terrain.GameMap;
+import Terrain.Location;
+import Terrain.TerrainType;
+import Units.Unit;
+import Units.UnitModel;
+import Units.Weapons.Weapon;
+
+/**
+ * Muriel will Make Units Reactively, Informed by the Enemy Loadout.
+ */
+public class Muriel implements AIController
+{
+  private Queue<GameAction> queuedActions = new ArrayDeque<GameAction>();
+
+  private Commander myCo = null;
+
+  private StringBuffer logger = new StringBuffer();
+  private int turnNum = 0;
+
+  private ArrayList<Commander> enemyCos = null;
+  
+  private UnitEffectivenessMap myUnitEffectMap;
+  private final double COST_EFFECTIVENESS_THRESHOLD = 0.75;
+
+  private ArrayList<XYCoord> nonAlliedProperties; // set from AIUtils.
+
+  public Muriel(Commander co, Commander[] allCos )
+  {
+    myCo = co;
+
+    // Initialize UnitModel collections.
+    ArrayList<UnitModel> myUnitModels = new ArrayList<UnitModel>();
+    enemyCos = new ArrayList<Commander>();
+    Map<Commander, ArrayList<UnitModel> > otherUnitModels = new HashMap<Commander, ArrayList<UnitModel> >();
+    for( Commander other : allCos )
+    {
+      if( myCo.isEnemy(other) )
+      {
+        enemyCos.add(other);
+        otherUnitModels.put(other, new ArrayList<UnitModel>());
+      }
+    }
+
+    // Figure out what I and everyone else can build.
+    for( UnitModel.UnitEnum modelEnum : UnitModel.UnitEnum.values() )
+    {
+      // Direct-fire UnitModels at my disposal.
+      UnitModel model = myCo.getUnitModel(modelEnum);
+      if( model != null )
+      {
+        myUnitModels.add(model);
+      }
+
+      // Direct-fire UnitModels for everyone else.
+      for( Commander oCo : enemyCos )
+      {
+        model = oCo.getUnitModel(modelEnum);
+        if( model != null )
+        {
+          otherUnitModels.get(oCo).add(model);
+        }
+      }
+    }
+
+    // Figure out unit matchups.
+    myUnitEffectMap = new UnitEffectivenessMap();
+    for( UnitModel myModel : myUnitModels )
+    {
+      for( Commander oCo : enemyCos )
+      {
+        for( UnitModel otherModel : otherUnitModels.get(oCo) )
+        {
+          Unit myUnit = new Unit(myCo, myModel);
+          Unit otherUnit = new Unit( oCo, otherModel );
+
+          double myDamage = 0;
+          Weapon myWeapon = myUnit.chooseWeapon(otherUnit.model, 1, false);
+          if( null != myWeapon )
+          {
+            BattleInstance.BattleParams params = new BattleInstance.BattleParams(myUnit, myWeapon,
+                otherUnit, Environment.getTile(TerrainType.ROAD, Weathers.CLEAR));
```

Comment:
It occurs to me that having the environment of the defender maybe isn't all that useful.
I guess a more comprehensive review is needed, but IDK where/when we'd have that discussion.

Reply:
Doesn't it need defender terrain to determine defense, which influences damage?

Reply:
...so it does. That could be easily passed in though, and is directly modifiable.

Reply:
You're right, we should make the `BattleParams` member variables private. I think I like keeping the terrain type in there, so the object is created with everything it needs to calculate the result of the battle.

Reply:
Nono, the whole point of BattleParams is to *make* things public.
The COs can then reach their dirty claws in there and change things to suit them, like certain COs could give an attack buff based on attack range or terrain.
I guess I should get around to pulling that stuff out of my AWBW branch and making PRs for here, heh.

Reply:
I feel like a lot of this design intent stuff should be documented somewhere, but I dunno where.

Reply:
Hm, yeah. OK. Comments are for documentation. Each class can have a javadoc block describing its use. Some do already. Unless you have a better idea, I'd stick with the simple no-frills approach.

Reply:
Hah, fair enough. I could complain that a lot of this stuff wasn't really set in stone yet, but that's not a very good excuse.